compio-io/src/read/mod.rs

```diff
@@ -0,0 +1,167 @@
+use std::{io::Cursor, rc::Rc, sync::Arc};
+
+use compio_buf::{BufResult, IoBuf, IoBufMut, IoVectoredBufMut, SetBufInit};
+
+mod buf;
+mod ext;
+
+pub use buf::*;
+pub use ext::*;
+
+use crate::util::slice_to_buf;
+
+/// AsyncRead
+///
+/// Async read with a ownership of a buffer
+pub trait AsyncRead {
+    /// Read some bytes from this source into the [`IoBufMut`] buffer and return
+    /// a [`BufResult`], consisting of the buffer and a [`usize`] indicating
+    /// how many bytes were read.
+    async fn read<B: IoBufMut>(&mut self, buf: B) -> BufResult<usize, B>;
+
+    /// Like `read`, except that it reads into a type implements
+    /// [`IoVectoredBufMut`].
+    async fn read_vectored<V: IoVectoredBufMut>(&mut self, buf: V) -> BufResult<usize, V>
+    where
+        V::Item: IoBufMut + SetBufInit;
+
+    /// Read the exact number of bytes required to fill the buf.
+    async fn read_exact<T: IoBufMut + SetBufInit>(&mut self, mut buf: T) -> BufResult<usize, T> {
+        let len = buf.buf_capacity() - buf.buf_len();
+        let mut read = 0;
+        while read < len {
+            let BufResult(result, buf_returned) = self.read(buf).await;
+            buf = buf_returned;
+            match result {
+                Ok(0) => {
+                    return BufResult(
+                        Err(std::io::Error::new(
+                            std::io::ErrorKind::UnexpectedEof,
+                            "failed to fill whole buffer",
+                        )),
+                        buf,
+                    );
+                }
+                Ok(n) => {
+                    read += n;
+                    unsafe { buf.set_buf_init(read) };
+                }
+                Err(ref e) if e.kind() == std::io::ErrorKind::Interrupted => {}
```

Comment:
I wonder if we will really receive EINTR... the driver will receive this indeed, but the runtime has dealt with it.

Reply:
Should I remove this line then?

Reply:
I think so. Actually you needn't write `match` manually. Instead, you can use `buf_try`. See the current `recv_exact` for reference.